File: Design-Patterns-Using-JavaScript/Bridge/BridgePatternShape.js
// Abstraction abstract class
class Shape {
    constructor(imp) {
        this.imp = imp;
    }

    draw() {
        throw new Error('Method draw() must be implemented');
    }
}

// Rectangle class
class Rectangle extends Shape {
    constructor(imp) {
        super(imp);
    }

    draw() {
        console.log(`Drawing Rectangle with colour ${this.imp.fill()}`);
    }
}

// Circle class
class Circle extends Shape {
    constructor(imp) {
        super(imp);
    }

    draw() {
        console.log(`Drawing Circle with colour ${this.imp.fill()}`);
    }
}

// Colour abstract class
class Colour {
    fill() {
        throw new Error('Method fill() must be implemented');
    }
}

// Red class
class Red extends Colour {
    fill() {
        return 'Red';
    }
}

// Green class
class Green extends Colour {
    fill() {
        return 'Green';
    }
}

// Blue class
class Blue extends Colour {
    fill() {
        return 'Blue';
    }
}

// Client code
const c1 = new Red();
let abstraction = new Circle(c1);
abstraction.draw();

const c2 = new Green();
abstraction = new Rectangle(c2);
abstraction.draw();



/*
Drawing Circle with color Red
Drawing Rectangle with color Green
*/

File: Design-Patterns-Using-Kotlin/Bridge/BridgePatternShape.kt
// Abstraction abstract class
abstract class Shape(protected val imp: Colour) {
    abstract fun draw()
}

// Rectangle class
class Rectangle(imp: Colour) : Shape(imp) {
    override fun draw() {
        println("Drawing Rectangle with colour ${imp.fill()}")
    }
}

// Circle class
class Circle(imp: Colour) : Shape(imp) {
    override fun draw() {
        println("Drawing Circle with colour ${imp.fill()}")
    }
}

// Colour interface
interface Colour {
    fun fill(): String
}

// Red class
class Red : Colour {
    override fun fill(): String {
        return "Red"
    }
}

// Green class
class Green : Colour {
    override fun fill(): String {
        return "Green"
    }
}

// Blue class
class Blue : Colour {
    override fun fill(): String {
        return "Blue"
    }
}

// Client code
fun main() {
    val c1: Colour = Red()
    var abstraction: Shape = Circle(c1)
    abstraction.draw()

    val c2: Colour = Green()
    abstraction = Rectangle(c2)
    abstraction.draw()
}

/*
Drawing Circle with color Red
Drawing Rectangle with color Green
*/
